### PR TITLE
Replace rm_log with leaf_set (compact bitmap)

### PR DESCRIPTION
**Pull request by antiochp**
#### New pmmr_leaf.bin file
Tracks all unpruned (unremoved) _leaf_ positions in the MMR as a compact (roaring) bitmap.
Self-prunes as positions are removed.
To facilitate rewind we store the bitmap of removed (spent outputs) positions per block in the db/index.
Rewinding a block involves rewinding inputs (by applying this bitmap) and rewinding outputs (by truncating the leaf_set at the appropriate output_pos (from the block_marker we are rewinding to).

#### Migration (local node rm_log -> leaf_set)
On startup we look for `rm_log.bin`. If we find one (and the leaf_set is empty) then we migrate to rm_log to new leaf_set, accounting for current chain state from both rm_log and prune_list.

#### Fast sync
The new `pmmr_leaf.bin` is not rewindable in a self-contained way. We need to know the positions of all inputs for all blocks being rewound (the old `rm_log.bin` _was_ rewindable as we stored block heights in the rm_log itself).
To allow fast-sync we now rewind _before_ sending the txhashset.zip across, taking a snapshot of the `pmmr_leaf.bin` file at the appropriate block height , and tagging the snapshot `pmmr_leaf.bin.xxxxxx` file with the block hash as filename suffix).
On receiving and validating the txhashset.zip file the receiving node copies the snapshot pmmr_leaf.bin.xxxxxx file across into the primary pmmr_leaf.bin file and uses this to validate the fast sync state).

TODO (migration & sync complicates things a bit) - 
- [x] fast sync (leaf_set -> leaf_set)
- [x] local node migration (rm_log -> leaf_set)
- [x] fast sync (rm_log -> leaf_set)

More TODO - 
- [x] rename `rewind_output_pos` and `rewind_spent_pos` to be more generic
    - [x] `rewind_add_pos` and `rewind_rm_pos` ?
    - [x] `UtxoSet` -> `LeafSet` also in grin_store crate?
- [ ] document/comment the rewind logic (what are these bitmaps we are passing around and why...
- [ ] does the lru_cache stuff address the I/O concerns fully?

---------------
[need to clean up the description here to reflect current reality]

We used to store the rm_log as a vec of u64 values, each with a corresponding index (block height) that the pos was removed.
This PR reworks this to store the rm_log as a compact bitmap, specifically a "roaring bitmap" -

https://roaringbitmap.org/
https://github.com/saulius/croaring-rs

We want the rm_log to be - 
* small (on disk and in memory)
* fast (to lookup pos in the rm_log)
* rewindable (so we can various validation scenarios in Grin, both block validation and tx validation) 

The current impl is not small (at least on a full archival node where we store _all_ removed pos, forever).
The current impl is also not particularly fast (via binary search) when the rm_log gets large.
The current impl _is_ good at rewinding (as we store the index on each entry in the rm_log) - any alternative also needs to be efficient at rewinding (to handle forks, build Merkle proofs, validate txs etc.)

TODO - 
- [x] work out efficient way of rewinding large number of pos (there must be a fast and clean way of truncating based on pos here)
- [x] cleanup check_compact to take pos (and not the index, or maybe both?)
- [x] check_compact is not fully implemented right now as we cannot call `removed_pre_cutoff()` on the bitmap impl (yet)
- [x] `croaring` only supports u32 values, do we _really_ need u64 for pos? (and if so there are documented ways of doing this, see below)
- [x] support backward compatibility/conversion 
    - [x] on startup read `pmmr_rm_log.bin` and migrate to `pmmr_utxo.bin`
- [x] do we want to use this for the prunelist also? Need to investigate further.

Related - #1137.
Started off looking at storing leaf pos only in the `rm_log` to keep it small and then took it a step further and stored the leaf pos in a bitmap rather than a big vec of positions.

### CODE REVIEW

**Comment by ignopeverell** [comment]
It'd be nice to have an idea of how it performs. Something tells me it _can't_ be slower than what we used to do, given that we were rewriting the whole thing every time, but just to get a basic idea especially around rewind.

A `u32` will get us to roughly 2.15 billion outputs ever created. Should be good for some time, but having a note about this somewhere in the code so we don't forget would also be good.

By the way is this true?

```
The current impl is not small (at least on a full archival node where we store all removed pos, forever).
```

If compaction doesn't work on archival nodes we need to take a look. Or do you mean that during sync? I think we should do intermediary compactions while syncing on archival nodes.

**Comment by antiochp** [comment]
@ignopeverell see chat in gitter but putting it here as well for reference.


I'm pretty sure "removing blocks from the index" and "pruning the MMRs" are tightly coupled right now.
Non-archival nodes do both of these every 2,000 blocks.
Archival nodes do _neither_ (and I have never seen a "prunelist" file on an archival node).

I'll open a separate issue to track this but it sounds like we want to decouple these two activities so they can be run separately without any dependencies.

Question though - if all nodes are pruning their local MMRs what happens if we encounter a reorg that goes earlier than the horizon, and no nodes on the network have an MMR that can be successfully rewound that far?
I guess archive nodes would need to rebuild MMR state from their db of blocks?

We trigger compacting here if non-archival, and compacting is for both the blocks in the db and the MMR pruning steps - 
https://github.com/mimblewimble/grin/blob/310c7b0b487da404944e56c45945537357c989a4/servers/src/common/adapters.rs#L476-L480

**Comment by ignopeverell** [comment]
Actually "pruning the MMR" also means 2 different things:

1. Cutting the data file here and there.
2. Compacting the rm log into a prune list (which may not be such a good name now).

I don't think archival nodes should do 1, but they probably should do 2. And the prune list, with a little more effort, should be rewindable to aby depth.

File: core/src/core/pmmr.rs [hunk 1]
```diff
@@ -410,67 +436,53 @@ where
...
-				current = parent;
-			} else {
-				break;
-			}
+		if self.backend.get_hash(position).is_none() {
+			return Ok(false);
 		}
-		self.backend.remove(to_prune, index)?;
+
+		self.backend.remove(position)?;
```
[the remark this answers is not in the document]
Comment (antiochp) [reply core/src/core/pmmr.rs]: Ahh good catch - it used to happen when we added pos to the rm_log.

Now this happens in `check_compact()`as part of the pruning logic - 
https://github.com/antiochp/grin/blob/ddeb3871430b3cfe6bb70e75316725125b0ca8ac/store/src/pmmr.rs#L368-L369

```
		let (leaves_removed, pos_to_rm) =
			self.pos_to_rm(cutoff_pos, rewind_output_pos, rewind_spent_pos);
```

Which calls `pos_to_rm()` which in turn expands the leaf nodes out into the set of pos to be pruned (accounting for previously pruned pos) - 
https://github.com/antiochp/grin/blob/ddeb3871430b3cfe6bb70e75316725125b0ca8ac/store/src/pmmr.rs#L468-L489

```
		for x in leaf_pos_to_rm.iter() {
			expanded.add(x);
			let mut current = x as u64;
			loop {
				let (parent, sibling) = family(current);
				let sibling_pruned = self.is_pruned(sibling);

				// if sibling previously pruned
				// push it back onto list of pos to remove
				// so we can remove it and traverse up to parent
				if sibling_pruned {
					expanded.add(sibling as u32);
				}

				if sibling_pruned || expanded.contains(sibling as u32) {
					expanded.add(parent as u32);
					current = parent;
				} else {
					break;
				}
			}
		}
```

File: core/src/core/pmmr.rs [hunk 2]
```diff
@@ -60,7 +63,12 @@ where
 	/// operations after that had been canceled. Expects a position in the PMMR
 	/// to rewind to as well as the consumer-provided index of when the change
 	/// occurred (see remove).
-	fn rewind(&mut self, position: u64, index: u32) -> Result<(), String>;
+	fn rewind(
+		&mut self,
+		position: u64,
+		rewind_output_pos: &Bitmap,
```
[the remark this answers is not in the document]
Comment (antiochp) [reply core/src/core/pmmr.rs]: Yeah agreed - will rename this to be more generic.

File: store/src/pmmr.rs [hunk 3]
```diff
@@ -197,25 +214,74 @@ where
...
 
+		let utxo_set_path = format!("{}/{}", data_dir, PMMR_UTXO_FILE);
+		let rm_log_path = format!("{}/{}", data_dir, PMMR_RM_LOG_FILE);
+
+		if let Some(header) = header {
+			let utxo_snapshot_path = format!("{}/{}.{}", data_dir, PMMR_UTXO_FILE, header.hash());
+			UtxoSet::copy_snapshot(utxo_set_path.clone(), utxo_snapshot_path.clone())?;
+		}
+
+		// If we need to migrate an old rm_log to a new utxo_set do it here before we
```
[the remark this answers is not in the document]
Comment (antiochp) [reply store/src/pmmr.rs]: Biggest concern here is the amount of "untested" changes going into testnet3.
I'd much rather migrate testnet2 and have this code tested in preparation for testnet3.

**Comment by antiochp** [comment]
One option with the IO cost may be to do something similar to the "header cache" that we added to the store. We keep the last n headers in memory in an LruCache to avoid needing to reread them from RocksDB repeatedly. We could easily do the same for n input bitmaps.

File: store/src/pmmr.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ignopeverell) [reply store/src/pmmr.rs]: :+1:

File: core/src/core/pmmr.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (antiochp) [reply core/src/core/pmmr.rs]: Renamed the following to remove all references to "spent" from grin_store crate.
This is now a lot more generic in terms of adding and removing _leaf_ positions (and not outputs).

`rewind_output_pos` -> `rewind_add_pos`
`rewind_spent_pos` -> `rewind_rm_pos`
`UtxoSet` -> `LeafSet`

**Comment by antiochp** [comment]
Merging onto master/testnet2.
If we need to iterate on this lets do it before testnet3.

**Pull request merged**